src/dataflow/src/server.rs

```diff
@@ -450,7 +466,11 @@ where
             }
             changes.compact();
             if !changes.is_empty() {
-                progress.push((id, changes));
+                progress.push(FrontierFeedback {
```

Comment:
If changes is empty but bindings isn't, the bindings will be ignored. This is a side effect of merging bindings into FrontierFeedback instead of making it a new message. The assert in the index branch of update_upper is another example of this type mismatch, where sometimes we expect something and sometimes we don't. I suggest exploring the idea of a separate message which might simplify the logic.

Reply:
ack. will try to do this